akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala

```diff
@@ -2209,69 +2209,72 @@ private[akka] final class StatefulMap[S, In, Out](
 
-      private var state: S = _
-      private var needInvokeOnCompleteCallback: Boolean = false
+      private var state: OptionVal[S] = _
 
       override def preStart(): Unit = {
-        state = create()
-        needInvokeOnCompleteCallback = true
+        state = OptionVal.Some(create())
       }
 
       override def onPush(): Unit =
         try {
           val elem = grab(in)
-          val (newState, newElem) = f(state, elem)
-          state = newState
+          val (newState, newElem) = f(state.get, elem)
+          state = OptionVal.Some(newState)
           push(out, newElem)
         } catch {
           case NonFatal(ex) =>
             decider(ex) match {
               case Supervision.Stop    => closeStateAndFail(ex)
               case Supervision.Resume  => pull(in)
-              case Supervision.Restart => resetStateAndPull()
+              case Supervision.Restart => restartState()
             }
         }
 
-      override def onUpstreamFinish(): Unit = closeStateAndComplete()
+      override def onUpstreamFinish(): Unit = {
+        completeStateIfNeeded() match {
+          case Some(elem) => emit(out, elem, () => completeStage())
+          case None       => completeStage()
+        }
+      }
 
       override def onUpstreamFailure(ex: Throwable): Unit = closeStateAndFail(ex)
 
       override def onDownstreamFinish(cause: Throwable): Unit = {
-        onComplete(state)
-        needInvokeOnCompleteCallback = false
+        completeStateIfNeeded()
         super.onDownstreamFinish(cause)
       }
 
-      private def resetStateAndPull(): Unit = {
-        needInvokeOnCompleteCallback = false
-        onComplete(state)
-        state = create()
-        needInvokeOnCompleteCallback = true;
-        pull(in)
-      }
-
-      private def closeStateAndComplete(): Unit = {
-        onComplete(state) match {
-          case Some(elem) => emit(out, elem, () => completeStage())
-          case None       => completeStage()
+      private def restartState(): Unit = {
+        completeStateIfNeeded() match {
+          case Some(elem) =>
+            emit(out, elem, () => state = OptionVal.Some(create()))
+          case None =>
+            state = OptionVal.Some(create())
+            // should always happen here but for good measure
+            if (!hasBeenPulled(in)) pull(in)
         }
-        needInvokeOnCompleteCallback = false
       }
 
       private def closeStateAndFail(ex: Throwable): Unit = {
-        needInvokeOnCompleteCallback = false
-        onComplete(state) match {
+        completeStateIfNeeded() match {
           case Some(elem) => emit(out, elem, () => failStage(ex))
           case None       => failStage(ex)
         }
       }
 
+      private def completeStateIfNeeded(): Option[Out] = {
+        state match {
+          case OptionVal.Some(s) =>
+            state = OptionVal.none[S]
+            onComplete(s)
```

Comment:
What if the state if `null`?

Reply:
That's true, but not sure if it makes a lot of sense to allow null as state?

Reply:
I think those two methods somekind of lifecycle methods, like prestart and poststop which will need always be call dispite the current state？

Reply:
Ok, fair enough, no way around a boolean flag then

Reply:
Maybe add some test to cover this? The operator is very tricky especially involve with state....
![image](https://user-images.githubusercontent.com/501740/192964558-c5dca489-73ec-4c3c-b159-53fafdb159f8.png)
- from roland.

Reply:
> Ok, fair enough, no way around a boolean flag then

I'm not sure I follow. It's lifecycle of the state. If the state has not been created `onComplete` should not be called, and it shouldn't be called more than once (and that is covered by resetting state to None.

I don't think we should allow `null` state and we can have a clear exception about that.

Reply:
The resource which can already be create or opened, and needs a clean up even a null returns, will a actor's postStop never called if the actor has a field which its current value is null?

And the state here cab be overrided with the returning state too.

Reply:
Add how can we implement fs2's mask operator if the onComplete will not be called if the state is null.

Reply:
You'd use option/optional/a collection to represent a state and the fact that it can be missing or present

Reply:
I'll keep this behavior and add docs/validation to fail if returning null

Reply:
@He-Pin , I'm a bit confused, it seems like there is nothing preventing you from just using `Option` instead of `null` like `def bracket[S, T](
      create: () => Option[S])(f: (Option[S], Out) => (S, T), onComplete: Option[S] => Option[T]): Repr[T]` to achieve the desired behavior.
And if that's absolutely needed, I would suggest create another variation that has `onComplete: Option[S] => Option[T]`.

Reply:
> @He-Pin , I'm a bit confused, it seems like there is nothing preventing you from just using `Option` instead of `null` like `def bracket[S, T](
>       create: () => Option[S])(f: (Option[S], Out) => (Option[S], T), onComplete: Option[S] => Option[T]): Repr[T]` to achieve the desired behavior.
> And if that's absolutely needed, I would suggest create another variation that has `onComplete: Option[S] => Option[T]`, it has more clear intension compare to relay on `null`.

That will add one more allocation.

Reply:
> I'll keep this behavior and add docs/validation to fail if returning null

I will take a try with cats effect's resource when I catch the plane. A keep of try finally , try close thing is essential from my perspective。

Reply:
For a case where the state can be present or not, and the allocation overhead of one per element is too much, a mutable wrapper class can be used. 

For the resource case, I don't understand why this is important, because if the resource is `null` there is nothing to close/clean up?

The "mask" operator, turning a failure to a complete is already covered by `recoverWith` + `Source.empty`, I don't see how that is relevant here.

Reply:
I think the oncomplete should works as  cats effects' Resource release, which will be called even the resource is null.

Then what about fs2's noneTerminate operator?

Reply:
I mean make the onComplete works as a finallizer thing. Going flightmode, see u guys.

Reply:
For the cats-effect/fs2 part, I think even though it's true that you **can** pass null into it, it is unlikely to be a design decision that you should rely on, a solid proof is if you search `null` in either cats-effect or fs2's website, there isn't any result. 
And even beyond the "practical reason", I would say FP people more dislike `null` than any other being, they wouldn't create an API that encourages you to use `null`.
The only valid argument I can see here is that it may be less appropriate to introduce a behavior change in a bugfix while a fix exists that doesn't change it.

Reply:
If you think about daliy java development, then that should be very valid:)